src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -120,118 +129,185 @@ module Make (Inputs : Inputs.S) :
+        Deferred.Or_error.fail @@ Error.of_string "disconnected chain"
+
+  let rec fold_until ~(init : 'accum)
+      ~(f :
+            'accum
+         -> 'a
+         -> ('accum, 'final) Continue_or_stop.t Deferred.Or_error.t)
+      ~(finish : 'accum -> 'final Deferred.Or_error.t) :
+      'a list -> 'final Deferred.Or_error.t = function
+    | [] ->
+        finish init
+    | x :: xs -> (
+        let open Deferred.Or_error.Let_syntax in
+        match%bind f init x with
+        | Continue_or_stop.Stop res ->
+            Deferred.Or_error.return res
+        | Continue_or_stop.Continue init ->
+            fold_until ~init ~f ~finish xs )
+
+  (* returns a list of state-hashes with the older ones at the front *)
+  let download_state_hashes ~logger ~trust_system ~network ~frontier ~num_peers
+      ~target_hash =
+    let peers = Network.random_peers network num_peers in
+    Logger.debug logger ~module_:__MODULE__ ~location:__LOC__
+      ~metadata:[("target_hash", State_hash.to_yojson target_hash)]
+      "Doing a catchup job with target $target_hash" ;
+    Deferred.Or_error.find_map_ok peers ~f:(fun peer ->
+        match%map
+          Network.get_transition_chain_witness network peer target_hash
+        with
+        | None ->
+            Or_error.errorf
+              !"Peer %{sexp:Network_peer.Peer.t} did not have the transition"
+              peer
+        | Some transition_chain_witness ->
+            let open Or_error.Let_syntax in
+            (* a list of state_hashes from new to old *)
+            let%bind hashes =
+              match
+                Transition_chain_witness.verify ~target_hash
+                  ~transition_chain_witness
+              with
+              | Some hashes ->
+                  return hashes
+              | None ->
+                  let error_msg =
+                    sprintf
+                      !"Peer %{sexp:Network_peer.Peer.t} sent us bad proof"
+                      peer
+                  in
+                  ignore
+                    Trust_system.(
+                      record trust_system logger peer.host
+                        Actions.(Violated_protocol, Some (error_msg, []))) ;
+                  Or_error.error_string error_msg
+            in
+            List.fold_until hashes ~init:[]
+              ~f:(fun acc hash ->
+                if Transition_frontier.find frontier hash |> Option.is_some
+                then Continue_or_stop.Stop (Ok (peer, acc))
+                else Continue_or_stop.Continue (hash :: acc) )
+              ~finish:(fun _ ->
+                Or_error.errorf
+                  !"Peer %{sexp:Network_peer.Peer.t} moves too fast"
+                  peer ) )
+
+  let verify_against_hashes transitions hashes =
+    List.length transitions = List.length hashes
+    && List.for_all2_exn transitions hashes ~f:(fun transition hash ->
+           State_hash.equal
+             ( External_transition.protocol_state transition
```

Comment:
NIT: can we make computing the state_hash of a transition be a function in the module, `External_transition`. The function is quite often.

Reply:
Agreed. I would do that.